CodeGen/Generators/QuantityRelationsParser.cs

```diff
@@ -0,0 +1,174 @@
+                .Select(r => r with { LeftQuantity = timeSpanQuantity })
+                .ToList());
+            relations.AddRange(relations
+                .Where(r => r.RightQuantity.Name is "Duration")
+                .Select(r => r with { RightQuantity = timeSpanQuantity })
+                .ToList());
+
+            // Sort all relations to keep generated operators in a consistent order.
+            relations.Sort();
+
+            foreach (var quantity in quantities)
+            {
+                var quantityRelations = new List<QuantityRelation>();
+
+                foreach (var relation in relations)
+                {
+                    if (relation.LeftQuantity == quantity)
+                    {
+                        // The left operand of a relation is responsible for generating the operator.
+                        quantityRelations.Add(relation);
+                    }
+                    else if (relation.RightQuantity == quantity && relation.LeftQuantity.Name is "double" or "TimeSpan")
+                    {
+                        // Because we cannot add generated operators to double or TimeSpan, we make the right operand responsible in this case.
+                        quantityRelations.Add(relation);
+                    }
+                }
+
+                quantity.Relations = quantityRelations.ToArray();
+            }
+        }
+
+        private static List<QuantityRelation> ParseRelations(string rootDir, IReadOnlyDictionary<string, Quantity> quantities)
+        {
+            var relationsFileName = Path.Combine(rootDir, "Common/UnitRelations.json");
+
+            try
+            {
+                var text = File.ReadAllText(relationsFileName);
+                var relationStrings = JsonConvert.DeserializeObject<List<string>>(text) ?? [];
+                relationStrings.Sort();
+
+                var parsedRelations = relationStrings.Select(relationString => ParseRelation(relationString, quantities)).ToList();
+
+                // File parsed successfully, save it back to disk in the sorted state.
+                File.WriteAllText(relationsFileName, JsonConvert.SerializeObject(relationStrings, Formatting.Indented));
+
+                return parsedRelations;
+            }
+            catch (Exception e)
+            {
+                throw new Exception($"Error parsing relations file: {relationsFileName}", e);
+            }
+        }
+
+        private static QuantityRelation ParseRelation(string relationString, IReadOnlyDictionary<string, Quantity> quantities)
+        {
+            var segments = relationString.Split(' ');
+
+            if (segments is not [_, "=", _, "*" or "/", _])
+            {
+                throw new Exception($"Invalid relation string: {relationString}");
+            }
+
+            var @operator = segments[3];
+            var left = segments[2].Split('.');
+            var right = segments[4].Split('.');
+            var result = segments[0].Split('.');
```

Comment:
This is fine, I was just thinking maybe regex was a good fit for this parsing?
Not important at all, just a tip to consider.

Reply:
I tried that but it wasn't a big improvement, the regex is quite ugly and less comprehensible than the list pattern:
```csharp
^(\w+)\.?(\w*) = (\w+)\.?(\w*) (\*|\/) (\w+)\.?(\w*)$

[_, "=", _, "*" or "/", _]
```